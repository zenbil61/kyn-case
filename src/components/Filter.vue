<template>
  <div class="filter-row">
    <label>{{ title }}</label>
    <div><slot></slot></div>
  </div>
</template>

<script>
export default {
  name: 'FilterComp',
  props: ['title'],
};
</script>

<style lang="scss">
.filter-row {
  border-radius: $radius;
  width: 100%;
  background: $gray;
  margin-bottom: 20px;
  padding: 20px;
  label {
    margin-bottom: 15px;
  }
}
</style>
