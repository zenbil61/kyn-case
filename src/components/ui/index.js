import Button from './Button.vue';
import Header from './Header.vue';
import HeaderItem from './HeaderItem.vue';
import NoResult from './NoResult.vue';
import Pagination from './Pagination.vue';
import PaginationItem from './PaginationItem.vue';
import Select from './Select.vue';

export { Button, Header, HeaderItem, NoResult, Pagination, PaginationItem, Select };
