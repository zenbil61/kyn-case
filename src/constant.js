const ROUTE = {
  PAGE_JOBS: '/jobs',
  PAGE_JOBS_DETAIL: '/jobs/:id',
};

const ACTIONS = {
  FETCH_JOB: 'fetchJob',
  FETCH_JOBS: 'fetchJobs',
  FETCH_CITIES: 'fetchCities',
};

const MUTATIONS = {
  SET_JOB: 'setJob',
  SET_JOB_LOADING: 'setJobLoading',
  SET_JOBS: 'setJobs',
  SET_JOBS_LOADING: 'setJobsLoading',
  SET_CITIES: 'setCities',
};

const CITIES = [
  { text: 'İstanbul', value: 34 },
  { text: 'Ankara', value: 6 },
  { text: 'İzmir', value: 35 },
];
export { ROUTE, ACTIONS, MUTATIONS, CITIES };
