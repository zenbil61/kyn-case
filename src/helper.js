function paginator(items, current_page, per_page_items) {
  let page = current_page || 1;
  let per_page = per_page_items || 10;
  let offset = (page - 1) * per_page;
  let paginatedItems = items.slice(offset).slice(0, per_page_items);
  return paginatedItems;
}
function wait(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
const debounce = (func, wait) => {
  let timeout;

  return function executedFunction(...args) {
    const later = () => {
      timeout = null;

      func(...args);
    };

    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  };
};

export { paginator, debounce, wait };
