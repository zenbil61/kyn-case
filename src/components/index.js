import Filter from './Filter.vue';
import Job from './Job.vue';

export { Filter, Job };
