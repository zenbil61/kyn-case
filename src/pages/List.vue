<template>
  <div class="row">
    <div class="col-md-3">
      <Filter title="Şehir Seçim.">
        <Select v-model="filter.city" :data="cities" />
      </Filter>

      <Filter title="Arama Yapın.">
        <input
          v-model="filter.search"
          type="text"
          class="form-control search-input"
          placeholder="Lütfen Arama Yapın.."
        />
      </Filter>

      <div class="btn-group filter-buttons">
        <Button :disabled="jobs_data_loading" data-test-id="search-btn" @click="handleSearch()">Ara</Button>
        <Button :disabled="jobs_data_loading" data-test-id="clear-btn" @click="handleClear()">Temizle</Button>
      </div>
    </div>
    <div class="col-md-9">
      <div v-if="jobs_data_loading">Yükleniyor...</div>
      <div v-if="!jobs_data_loading">
        <div class="filter-detail">
          <h1>Toplam {{ jobs_data?.totalCount }} adet ilan bulundu..</h1>

          <p v-if="route.query.search || route.query.city">
            <span v-if="route.query.city">"{{ cityName }}" şehrindeki </span>
            <span v-if="route.query.search"> "{{ route.query.search }}" </span>
            arama sonucu
          </p>
        </div>

        <NoResult v-if="!jobs_data?.jobs?.length && !jobs_data_loading" />

        <Job
          v-for="item in jobs_data.jobs"
          :key="item.id"
          :title="item.title"
          :description="item.description"
          @click="handleDetail(item.id)"
        />
        <Pagination v-if="jobs_data?.totalPage > 1">
          <PaginationItem @click="handlePage('prev')">Önceki</PaginationItem>
          <PaginationItem v-for="i in jobs_data.totalPage" :key="i" :active="i === filter.page" @click="handlePage(i)">
            {{ i }}
          </PaginationItem>
          <PaginationItem @click="handlePage('next')">Sonraki</PaginationItem>
        </Pagination>
      </div>
    </div>
  </div>
</template>
<script>
export default {
  name: 'ListComp',
};
</script>
<script setup>
import { ref, computed, onMounted } from 'vue';
import { Button, Select, Pagination, PaginationItem, NoResult } from '../components/ui';
import { Job, Filter } from '../components';
import { ROUTE, ACTIONS } from '../constant';
import { useRouter, useRoute } from 'vue-router';
import { useStore } from 'vuex';

const router = useRouter();
const route = useRoute();
const store = useStore();

let filter = ref({
  search: route.query.search,
  city: route.query.city,
  page: parseInt(route.query.page) || 1,
});

// computed
const jobs_data = computed(() => store.state.jobs_data);
const jobs_data_loading = computed(() => store.state.jobs_data_loading);
const cityName = computed(() => store.getters.cityName(route.query.city));
const cities = computed(() => store.state.cities);

// events
const handleDetail = (id) => {
  router.push(ROUTE.PAGE_JOBS_DETAIL.replace(':id', id));
};
const handleClear = () => {
  router.replace({ query: null }).then((resp) => {
    filter.value = {
      page: 1,
    };
    getJobs();
  });
};
const handleSearch = () => {
  const isSearchChange = route.query.search !== filter.value.search;
  const isCityChange = route.query.city !== filter.value.city;

  if (isSearchChange || isCityChange) {
    filter.value.page = 1;
  }

  router.replace({ query: filter.value }).then((resp) => getJobs());
};
const handlePage = (page) => {
  const oldPage = filter.value.page;
  const isLastPage = filter.value.page >= jobs_data.value.totalPage;
  const isFirstPage = filter.value.page <= 1;

  if (page === 'next' && !isLastPage) {
    filter.value.page++;
  } else if (page === 'prev' && !isFirstPage) {
    filter.value.page--;
  } else if (typeof page === 'number') {
    filter.value.page = page;
  }
  // aynı sayfaya tıklarsa fazla request atmasın kontrolü
  if (filter.value.page !== oldPage) handleSearch();
};

const getJobs = () => {
  store.dispatch(ACTIONS.FETCH_JOBS, {
    ...filter.value,
  });
};

onMounted(() => {
  if (!store.state.cities || store.state.cities.length < 1) {
    router.push('/');
    return;
  }
  getJobs();
});
</script>

<style lang="scss">
.filter-buttons {
  width: 100%;
  button {
    width: 50%;
  }
}
.filter-detail {
  p {
    color: $light-color;
  }
}
</style>
