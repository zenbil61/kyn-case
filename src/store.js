import { createStore } from 'vuex';
import { ACTIONS, MUTATIONS } from './constant';
import axios from 'axios';
import { wait } from './helper';
const BASE_API_URL = process.env.VUE_APP_BASE_API_URL;

const store = createStore({
  state() {
    return {
      jobs_data: {},
      jobs_data_loading: true,
      cities: [],
      job: {},
      job_loading: true,
    };
  },
  mutations: {
    [MUTATIONS.SET_JOBS]: (state, data) => (state.jobs_data = data),
    [MUTATIONS.SET_JOBS_LOADING]: (state, data) => (state.jobs_data_loading = data),
    [MUTATIONS.SET_JOB]: (state, data) => (state.job = data),
    [MUTATIONS.SET_JOB_LOADING]: (state, data) => (state.job_loading = data),
    [MUTATIONS.SET_CITIES]: (state, data) => (state.cities = data),
  },
  actions: {
    async [ACTIONS.FETCH_JOBS]({ commit }, request) {
      commit(MUTATIONS.SET_JOBS_LOADING, true);
      await wait(500);
      const { data } = await axios.get(`${BASE_API_URL}/jobs`, {
        params: request,
      });
      commit(MUTATIONS.SET_JOBS, data.data);
      commit(MUTATIONS.SET_JOBS_LOADING, false);
    },
    async [ACTIONS.FETCH_CITIES]({ commit }) {
      const { data } = await axios.get(`${BASE_API_URL}/cities`);
      commit(MUTATIONS.SET_CITIES, data.data);
    },
    async [ACTIONS.FETCH_JOB]({ commit }, { id }) {
      commit(MUTATIONS.SET_JOB_LOADING, true);
      await wait(500);

      const { data } = await axios.get(`${BASE_API_URL}/jobs/detail`, {
        params: { id },
      });
      commit(MUTATIONS.SET_JOB, data.data);
      commit(MUTATIONS.SET_JOB_LOADING, false);
    },
  },
  getters: {
    cityName: (state) => (id) => {
      return state.cities.find((city) => city.value == id).text;
    },
  },
});

export default store;
