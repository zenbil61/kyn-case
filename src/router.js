import { createRouter, createWebHistory } from 'vue-router';

import Home from './pages/Home.vue';
import Detail from './pages/Detail.vue';
import List from './pages/List.vue';

import { ROUTE } from './constant';
const routes = [
  { path: '/', component: Home },
  { path: ROUTE.PAGE_JOBS, component: List },
  {
    path: ROUTE.PAGE_JOBS_DETAIL,
    component: Detail,
  },
];
const router = createRouter({
  history: createWebHistory(),
  routes,
});

export default router;
