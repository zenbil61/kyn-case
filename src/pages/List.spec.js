import { mount } from '@vue/test-utils';
import List from './List.vue';
import router from '@/router';
import store from '@/store';
import { wait } from '@/helper';
import { CITIES } from '@/constant';
import { PaginationItem, Pagination } from '@/components/ui';

it('renders correctly', async () => {
  store.state.cities = CITIES;
  const wrapper = mount(List, {
    global: {
      plugins: [router, store],
    },
  });
  await wait(1500);
  expect(wrapper.element).toMatchSnapshot();

  // pagination control
  const pagination = wrapper.findComponent(Pagination);
  expect(pagination.exists()).toBeTruthy();

  // pagination click control
  pagination.findAllComponents(PaginationItem).at(2).trigger('click');
  await wait(1500);
  expect(wrapper.element).toMatchSnapshot();

  // search control
  const textInput = wrapper.find('input[type="text"]');
  await textInput.setValue('front');
  const select = wrapper.find('select');
  await select.setValue(34);
  const searchButton = wrapper.find('[data-test-id=search-btn]');
  await searchButton.trigger('click');
  await wait(1500);
  expect(wrapper.element).toMatchSnapshot();
});
