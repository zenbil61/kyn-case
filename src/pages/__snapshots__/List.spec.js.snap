// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="row"
>
  <div
    class="col-md-3"
  >
    <div
      class="filter-row"
    >
      <label>
        Şehir Seçim.
      </label>
      <div>
        
        <select
          class="form-select"
        >
          <option
            disabled=""
            value=""
          >
            Seçim Yapın
          </option>
          
          <option
            value="34"
          >
            İstanbul
          </option>
          <option
            value="6"
          >
            Ankara
          </option>
          <option
            value="35"
          >
            İzmir
          </option>
          
        </select>
        
      </div>
    </div>
    <div
      class="filter-row"
    >
      <label>
        Arama Yapın.
      </label>
      <div>
        
        <input
          class="form-control search-input"
          placeholder="Lütfen Arama Yapın.."
          type="text"
        />
        
      </div>
    </div>
    <div
      class="btn-group filter-buttons"
    >
      <button
        class="btn btn-primary"
        data-test-id="search-btn"
        type="button"
      >
        
        Ara
        
      </button>
      <button
        class="btn btn-primary"
        data-test-id="clear-btn"
        type="button"
      >
        
        Temizle
        
      </button>
    </div>
  </div>
  <div
    class="col-md-9"
  >
    <!--v-if-->
    <div>
      <div
        class="filter-detail"
      >
        <h1>
          Toplam 9 adet ilan bulundu..
        </h1>
        <!--v-if-->
      </div>
      <!--v-if-->
      
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            frontend developer 1
          </h5>
          <p
            class="card-text"
          >
            İstanbul üsküdar adresinde tam zamanlı çalışacak frontend developer arayışımız devam etmektedir
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Backend developer 1
          </h5>
          <p
            class="card-text"
          >
            Ankara mamak adresinde  parttime(remote) çalışacak backend developer arayışımız devam etmektedir
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            mobile developer 1
          </h5>
          <p
            class="card-text"
          >
            İstanbul fatih lokasyonunda hibrit olarak haftada 2 gün ofis 3 gün remote şeklinde çalışacak android developer aranmaktadır
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      
      <nav
        class="job-pagination"
      >
        <ul
          class="pagination"
        >
          
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              Önceki
              
            </a>
          </li>
          
          <li
            active="true"
            class="page-item active"
          >
            <a
              class="page-link"
              href="#"
            >
              
              1
              
            </a>
          </li>
          <li
            active="false"
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              2
              
            </a>
          </li>
          <li
            active="false"
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              3
              
            </a>
          </li>
          
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              Sonraki
              
            </a>
          </li>
          
        </ul>
      </nav>
    </div>
  </div>
</div>
`;

exports[`renders correctly 2`] = `
<div
  class="row"
>
  <div
    class="col-md-3"
  >
    <div
      class="filter-row"
    >
      <label>
        Şehir Seçim.
      </label>
      <div>
        
        <select
          class="form-select"
        >
          <option
            disabled=""
            value=""
          >
            Seçim Yapın
          </option>
          
          <option
            value="34"
          >
            İstanbul
          </option>
          <option
            value="6"
          >
            Ankara
          </option>
          <option
            value="35"
          >
            İzmir
          </option>
          
        </select>
        
      </div>
    </div>
    <div
      class="filter-row"
    >
      <label>
        Arama Yapın.
      </label>
      <div>
        
        <input
          class="form-control search-input"
          placeholder="Lütfen Arama Yapın.."
          type="text"
        />
        
      </div>
    </div>
    <div
      class="btn-group filter-buttons"
    >
      <button
        class="btn btn-primary"
        data-test-id="search-btn"
        type="button"
      >
        
        Ara
        
      </button>
      <button
        class="btn btn-primary"
        data-test-id="clear-btn"
        type="button"
      >
        
        Temizle
        
      </button>
    </div>
  </div>
  <div
    class="col-md-9"
  >
    <!--v-if-->
    <div>
      <div
        class="filter-detail"
      >
        <h1>
          Toplam 9 adet ilan bulundu..
        </h1>
        <!--v-if-->
      </div>
      <!--v-if-->
      
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            frontend developer 2
          </h5>
          <p
            class="card-text"
          >
            İstanbul üsküdar adresinde tam zamanlı çalışacak frontend developer arayışımız devam etmektedir
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Backend developer 2
          </h5>
          <p
            class="card-text"
          >
            Ankara mamak adresinde  parttime(remote) çalışacak backend developer arayışımız devam etmektedir
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            mobile developer 2
          </h5>
          <p
            class="card-text"
          >
            İstanbul fatih lokasyonunda hibrit olarak haftada 2 gün ofis 3 gün remote şeklinde çalışacak android developer aranmaktadır
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      
      <nav
        class="job-pagination"
      >
        <ul
          class="pagination"
        >
          
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              Önceki
              
            </a>
          </li>
          
          <li
            active="false"
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              1
              
            </a>
          </li>
          <li
            active="true"
            class="page-item active"
          >
            <a
              class="page-link"
              href="#"
            >
              
              2
              
            </a>
          </li>
          <li
            active="false"
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              3
              
            </a>
          </li>
          
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              
              Sonraki
              
            </a>
          </li>
          
        </ul>
      </nav>
    </div>
  </div>
</div>
`;

exports[`renders correctly 3`] = `
<div
  class="row"
>
  <div
    class="col-md-3"
  >
    <div
      class="filter-row"
    >
      <label>
        Şehir Seçim.
      </label>
      <div>
        
        <select
          class="form-select"
        >
          <option
            disabled=""
            value=""
          >
            Seçim Yapın
          </option>
          
          <option
            value="34"
          >
            İstanbul
          </option>
          <option
            value="6"
          >
            Ankara
          </option>
          <option
            value="35"
          >
            İzmir
          </option>
          
        </select>
        
      </div>
    </div>
    <div
      class="filter-row"
    >
      <label>
        Arama Yapın.
      </label>
      <div>
        
        <input
          class="form-control search-input"
          placeholder="Lütfen Arama Yapın.."
          type="text"
        />
        
      </div>
    </div>
    <div
      class="btn-group filter-buttons"
    >
      <button
        class="btn btn-primary"
        data-test-id="search-btn"
        type="button"
      >
        
        Ara
        
      </button>
      <button
        class="btn btn-primary"
        data-test-id="clear-btn"
        type="button"
      >
        
        Temizle
        
      </button>
    </div>
  </div>
  <div
    class="col-md-9"
  >
    <!--v-if-->
    <div>
      <div
        class="filter-detail"
      >
        <h1>
          Toplam 3 adet ilan bulundu..
        </h1>
        <p>
          <!--v-if-->
          <span>
             "front" 
          </span>
           arama sonucu 
        </p>
      </div>
      <!--v-if-->
      
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            frontend developer 1
          </h5>
          <p
            class="card-text"
          >
            İstanbul üsküdar adresinde tam zamanlı çalışacak frontend developer arayışımız devam etmektedir
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            frontend developer 2
          </h5>
          <p
            class="card-text"
          >
            İstanbul üsküdar adresinde tam zamanlı çalışacak frontend developer arayışımız devam etmektedir
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      <div
        class="card job"
      >
        <img
          class="card-img-top"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            frontend developer 3
          </h5>
          <p
            class="card-text"
          >
            İstanbul üsküdar adresinde tam zamanlı çalışacak frontend developer arayışımız devam etmektedir
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            
            Detay
            
          </button>
        </div>
      </div>
      
      <!--v-if-->
    </div>
  </div>
</div>
`;
